_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,223 @@
+---
+layout: post
+title: A story unfolds
+category: Haskell
+description: or how I tried to plant a tree
+---
+# A story unfolds
+
+### or how I tried to plant a tree
+
+I tried to re-implement the Unix `tree` tool in Haskell, as a learning exercise. The goal is to create a tool that, when provided with a directory, lists all the contents and prints them in a way similar to this:
+```
+$ tree
+.
+├── a
+│   └── c
+└── b
+    ├── d
+    └── e
+
+5 directories, 0 files
+```
+It is well-defined, well-structured, and obviously recursive problem, so it should be easy to do, a no-brainer almost[^excuses].
+
+I started hacking away at the tree.
+
+The thing I would need first was a function to list directory contents, and another one to let me distinguish between directories and files. Quite fortunately, I had a pair of trusty axes on hand[^axes]:
+```haskell
+listDir :: File -> IO [File]
+isDir :: File -> IO Boolean
+```
+A naïve implementation came to mind right away[^madeup]:
+```haskell    
+tree :: File -> IO ()
+tree file = tree' "" file
+  where tree' prefix file = do
```

Comment:
That's effectively an inner function within `tree` right?

Reply:
Yes.